workspaces/src/cargo/mod.rs

```diff
@@ -0,0 +1,101 @@
+use anyhow::anyhow;
+use async_process::Command;
+use cargo_metadata::{Message, Metadata, MetadataCommand};
+use std::env;
+use std::fmt::Debug;
+use std::fs;
+use std::path::Path;
+use std::process::Stdio;
+use tracing::debug;
+
+fn cargo_bin() -> Command {
+    match env::var_os("CARGO") {
+        Some(cargo) => Command::new(cargo),
+        None => Command::new("cargo"),
+    }
+}
+
+fn cargo_metadata<P: AsRef<Path> + Debug>(project_path: P) -> anyhow::Result<Metadata> {
+    let mut metadata_command = MetadataCommand::new();
+    metadata_command.current_dir(project_path.as_ref());
```

Comment:
@itegulov, so this actually wasn't what I meant either 😅.. How about you try omitting this line and not setting `current_dir`, so when `cargo` builds the contract, it'll reuse the target directory of the project. Instead of creating a separate one inside the contract directory.

Reply:
Well, `cargo metadata` should not build anything anyway, right? If I get rid of `current_dir`, then how would `CargoCommmand` know where the directory containing the contract is located at?

Reply:
`cargo build` already knows that from here: https://github.com/near/workspaces-rs/blob/eff8274fc35418b668d59b29887b601b2112f60f/workspaces/src/cargo/mod.rs#L38

The point of not setting the `current_dir` for `cargo metadata` is so we can get the `target` directory of the parent.

So, while `cargo` builds in the directory containing the contract, it keeps all the build artifacts in the same `target` directory.

Reply:
For example, in the case of the `test-contract-status-message` contract, we have a tree like this:

```diff
 near/workspaces-rs
 ├── target
+│   └── wasm32-unknown-unknown
+│       └── release
+│           └── test_contract_status_message.wasm
 └── workspaces
     └── tests
         └── test-contracts
             └── status-message
-                └── target
-                    └── wasm32-unknown-unknown
-                        └── release
-                            └── test_contract_status_message.wasm
```

Reply:
> The point of not setting the current_dir for cargo metadata is so we can get the target directory of the parent.

Exactly, when you do `cargo metadata` in one of the workspaces, it returns `target_directory` that is equal to the target directory of the parent.

> So, while cargo builds in the directory containing the contract, it keeps all the build artifacts in the same target directory.

And to ensure that it works as you describe it I set `--target-dir` to the output of `cargo metadata` from the previous paragraph.

Sorry if I am speaking nonsense here, but I am really confused how is my approach is different to what you are suggesting 😅

Reply:
Hmm, but `test-contract-status-message` is not a workspace of the root cargo project, but rather an independent cargo project. Would it be safe to reuse root target directory for building it?

Reply:
I'd say its pretty safe to put it in the root target dir. Pretty much everything workspaces-rs generates goes in there including the downloaded sandbox binary. It's also a very clean way of later deleting everything with just a `cargo clean` at the root level

Reply:
Ok, I have tested this and indeed it seems to work fine. I could find any definitive information on "safeness" to do so, but I guess since the feature is unstable anyway we can try and see if it works as expected for users?